matchit / matchparen Lua implementation [skip ci]

yochem opened the pull request:
Minimal Lua version of Vim's matchit plugin that is treesitter-aware but supports syntax as fallback.

*WIP*

## TODO

- [ ] Jump to unmatched group `]%` `[%`
- [ ] Select surrounding group `v_a%`
- [ ] Treesitter: Support middle keywords ('elseif' and friends)
- [ ] Treesitter: HTML tags (everything not keyword or bracket)
- [ ] Syntax: Wrap around after end of pair
- [ ] Tests

I want to add:
- [ ] Text object `i%` and `a%` (currently only `%` which acts like `w` with words)

## Deprecated
- matchparen:
  - `:DoMatchParen` / `:NoMatchParen`
  - `:let matchparen_disable_cursor_hl = 1`
- matchit:
  - `:MatchDisable` / `:MatchEnable`
  - `b:match_words` backrefs higher than `\1`

## New features

- Treesitter support
- Highlight keyword group under the cursor (like matchparen does for brackets) (opt-in?)
- Ternary operator support


---

References:
- https://github.com/neovim/neovim/issues/15462
- https://github.com/neovim/neovim/pull/31563 
- https://github.com/neovim/neovim/issues/15462#issuecomment-2481544422

yochem committed:
Let's figure this out

clason commented:
Don't forget `matchparen.vim` -- these are so similar that they should be able to share a lot of code.

yochem commented:
Ah yes, forgot to mention it. That's also the idea indeed!

On that note, should this plugin be an almost compatible replacement for both, or just replace the common functionality but in another way?

For example, should we keep things like the matchit/matchparen names, `:DoMatchParen`, `:NoMatchParen`, pack/dist/opt/ path, etc?

clason commented:
It should be a functional replacement, but no need to replicate the full interface (unless it's natural).

And just for the record (I understand that this is draft and still early testing): The final module should be exposed similarly to `comment.lua`.

lewis6991 commented:
I don't know if it's useful but I also implemented something very similar to this with https://github.com/lewis6991/nvim-treesitter-pairs . Does `@keyword` and `@punctuation.bracket`

yochem commented:
> It should be a functional replacement, but no need to replicate the full interface (unless it's natural).

Noted!

> And just for the record (I understand that this is draft and still early testing): The final module should be exposed similarly to comment.lua.

Thanks for this. Even though this is indeed still a very early draft, this kind of feedback is already very welcome.

> I don't know if it's useful but I also implemented something very similar to this with https://github.com/lewis6991/nvim-treesitter-pairs . Does @keyword and @punctuation.bracket

Thanks @lewis6991, indeed very similar, and definitely useful!

yochem committed:
Add early start of matchparen

yochem committed:
matchparen: syntax support

yochem committed:
Type annotations

clason commented on runtime/lua/vim/_matchpairs.lua:
```diff
@@ -20,6 +20,13 @@ local function jump_to_node(node, start)
 end
 
+--- @param node TSNode
+local function anonymous_children(node)
```
Note for later: consider `vim._,memoize` for potentially expensive treesitter calculations.

clason commented:
> Limit search: visible in window, 100 lines above/below, 'synmaxcol', timeout?

Ideally, the Lua rewrite (together with treesitter performance improvements in progress) will remove any performance concerns, so this shouldn't be necessary (and hence configurable).

yochem committed:
matchparen: support skipping syntax groups

yochem committed:
matchparen: Fix searchskip

clason commented:
You can use Vim regex from Lua. If we memoize the parsed `match_words` per buffer, that should not be a performance bottleneck.

yochem commented:
Thanks, I'll try t do that more. The problem is that there is a lot of manual parsing involved, e.g. manually finding capture groups, and then swapping the backrefs with the found groups. Just wish they would've supported a stricter and simpler pattern for the matching words 😛.

dsully commented:
I also implemented a version of this using tree-sitter which handles the matching functionality:

https://github.com/dsully/treesitter-jump.nvim

yochem commented:
Thanks! I'll definitely have a look at it. However, one of the goals is to have a language agnostic implementation that does not require new TS queries / captures, and instead reuse the those from the TS highlights (which are ofc not language agnostic)

yochem committed:
Copy resolve from matchit.vim

yochem committed:
Change namespace/augroup format

See #32001

lewis6991 commented:
> As I said earlier, the b:match_words format is quite extensive and complex. 

My vision for this was that we would eventually adapt these into equivalent treesitter queries, either as a separate file or bundled into `highlights.scm`. I don't want the `b:match_words` concept to live on indefinitely. We should move forward to query files which are much more precise and structured.

clason commented:
Well, that would require maintaining a parallel universe (in nvim-treesitter, so I have some standing on this issue), while we get `b:match_words` for free from Vim's ftplugins. So we shouldn't throw that away on a whim.

Of course,
1. we should only implement actually the minimal format used in our runtime,
2. we can still add "pair queries" if that opens up sufficient improvements (which I don't see right now) to merit dropping filetypes without tree-sitter parser and rewriting the implementation for those with.

This does not imply a commitment to keep these around forever; I just think the bar for pulling that trigger is relatively high (given current understanding). In any case, this discussion is not a blocker for this PR; I would just prefer to have _actual_ feature parity with `matchit` if possible (i.e., no significant out-of-the-box regression; I don't care about hypothetical features the format supports).

yochem committed:
More work on matchit:syntax

yochem committed:
Rename vim._matchpairs to vim._matching and set default maps/autocmds

yochem committed:
Combine implementations for highlighting and jumping

yochem committed:
Support [count] for % and g%

yochem committed:
Add Matching documentation

yochem commented:
I combined the implementations for highlighting pairs and jumping between groups. Now you get highlighting keywords "for free". I don't think that many people want this, so it shouldn't be the default, but could be nice to be able to configure it. For example:

```lua
-- user can add this to config instead:
vim.api.nvim_create_autocmd({ 'CursorMoved', 'CursorMovedI', 'WinEnter' }, {
  callback = vim.schedule_wrap(function()
    require('vim._matching').highlight({
      keywords = true,
    })
  end),
})
```

Or maybe this could be something that is exposed in `vim.hl`? What do you think?

(Not that important and could always be added later, but I noticed it as I was currently working on it.)

yochem committed:
Move matchparen and matchit to pack/dist/opt

dsully commented on runtime/doc/various.txt:
```diff
@@ -626,4 +626,62 @@ gc			Text object for the largest contiguous block of
+|Funcref| as value: >vim
+	:let b:match_words = 'GetMatchWords()'
+<
+The group definitions also support a single backref group |\1|. For example,
+>vim
+	:let b:match_words = '\<\(if|function\)\>:\<end\1\>'
+<
+matches "if" with "endif" and "function" with "endfunction".
```
Should there be a way to set this for multiple file types at a time? Otherwise people will need a bunch of `after/ftplugin/<lang>.lua` files to set `match_words`

Something more like how `vim.filetype.add()` is implemented?

clason replied:
That is simply for backward compatibility with existing ftplugins. Going forward, it will not be advertised as a first-class configuration interface. (That is still TBD.)

yochem replied:
Should I keep it in the documentation? I wanted to keep the `b:match_words` tag for now.

clason replied:
Yes, but it should state explicitly that these are supported for the sake of compatibility with `matchit`. Since we will keep shipping `matchit` (just not enable it by default), we can just link to their documentation and focus on our limitations (if any).

yochem committed:
Update docs

yochem committed:
docs: deprecation notice for matchit and matchparen

yochem committed:
Support b:match_ignorecase

yochem committed:
fix(matching): [count]% should work as |N%|

yochem committed:
fix: nvim_buf_add_highlight is deprecated in #32099

yochem committed:
feat(matching): Prefer matchparen/matchit if they're in runtimepath

Problem: No way to keep using old optional plugins matchparen / matchit.

Solution: If found in rtp, don't run matching functionality. It is also
possible to load only one (e.g. only matchparen).

yochem commented:
Thanks for the interest and your implementation @monkoose! I haven't worked on this PR because, like you mentioned, fully covering matchit, especially with treesitter, is quite hard and has a lot of edge cases to cover. I was actually thinking about asking if it would be okay to first merge a matchparen alternative and later extend it with matchit functionality. 

The parts I struggled with (from the top of my head, haven't looked at it for a while):
- unmatched groups (an `else` without `if`)
- 'middle' parts of groups (`elif`)
- including things like html tags without adding TS capture groups
- user configurable matching groups using TS queries. Maybe better for s follow-up PR, but I think adding custom match-groups using TS queries is much better than the current Vim 'matchpairs'. 

I don't think I'll work on the full implementation of machit + matchparen for now, **so if anyone wants to take over the work, go for it**!

I'm available to review/discuss continued work on this. Furthermore, if core would consider to accept a PR that for now only implements matchparen behaviour, I could create a new PR for that. Let me know.

yochem commented:
> There is no debounce (like in default matchparen), so in some cases it could visually slowdown cursor movement in some cases, especially in regex syntax buffers with heavy syntax logic (where getting synID is slow).

Yeah I hadn't spend to much time on the syntax implementation (basically tried to convert the vimscript version to lua), as I thought the TS version would be used more anyway (and I liked working on that more). The syntax implementation was more of a fallback.

Thank you for your detailed analysis and comparison :)

clason commented:
The main goal here is still the matchit/matchparen integration, but the matchit part doesn't have to be perfect. Things like middle delimiters and custom queries can absolutely be left for follow up PRs. 

(We still ship matchit as an optional plugin after all so that users can enable it if they see unacceptable regressions.)

yochem committed:
Move to only matchparen for now

TODO: revert (doc) edits for matchit

ofseed commented:
It looks like this PR has been stuck for a bit. Would you mind if I also gave it a try? I'm planning to start with a Lua rewrite that stays as consistent as possible with the current version. It will retain but reduce the use of Vimscript expressions, and it won't involve any tree-sitter changes, so the approach will be quite different from this PR. I think this might be a good starting point for us.

yochem commented:
Yes please, go for it! I haven't had the time to revisit this and dive deeper in the nitty gritty of matchit and how to replace that with TS. Happy to review your work when it's ready.

ofseed commented:
I'm starting with something simple first by rewriting `matchparen`, see #39625.

Follow-ups are tracking in #39683.

monkoose commented:
Did some battle tests of multiple matchparen implementations (vimscript matchparen, new merged lua matchparen, this PR, [vim-matchup](https://github.com/andymass/vim-matchup) and mine plugin [matchparen.nvim](https://github.com/monkoose/matchparen.nvim))

I'm using this files for it

 https://gist.github.com/monkoose/24dbced1234092c7bf186fd2f6875d69 (file with the long line)

https://gist.github.com/monkoose/284f02578631d30c487ec239a9f834c7 (file with a lot of curly braces that can be visible on the screen)

Using D language because it has on of the heaviest treesitter parser in my experience.
To be fair setup `matchparen.nvim` with `setup({debounce = 0})` to disable debouncing. (it still has the difference of hardcoded search limit to 150ms and matchparen has configurable option of 300ms by default) but it doesn't really matter for this tests.

For `curly_braces.d` file the best test line is 27, 31 and string lines. For 27 and 31 lines put the cursor at the start of it and just hold `l` and see how fast it starts lagging or stops highlighting.
On my PC starting from the worst:

vimscript matchparen (default options):
- syntax off (line 31): starts lagging immediately.
- regex syntax (line 27): starts lagging immediately. Highlights wrong brackets inside the string (line 24 and 28).
- treesitter (line 31): start lagging immediately and skips highlights. Highlights wrong brackets inside the string.
Bottlenecks: vimscript, `synstack()`, `searchpair()`, wrong highlights.

this PR (default options):
- syntax off (line 31): starts lagging at the quarter of the line.
- regex syntax (line 27): starts lagging almost immediately. Highlights wrong brackets inside the string.
- treesitter (line 31): start lagging immediately Highlights wrong brackets inside the string.
Bottlenecks: `synstack()`, `searchpair()`, wrong highlights.

new lua merged matchparen (default options):
- syntax off (line 31): starts lagging immediately.
- regex syntax (line 27): starts lagging at the quarter of the line. Highlights wrong brackets inside the string.
- treesitter (line 31): start lagging immediately and skips highlights. Highlights wrong brackets inside the string.
Bottlenecks: `synstack()`, `searchpair()`, wrong highlights.

vim-matchup (default options):
- syntax off (line 31): starts lagging immediately.
- regex syntax (line 27): starts lagging around the half of the line. Right highlights inside strings.
- treesitter (line 31): start lagging immediately and skips highlights. Highlights wrong brackets inside the string, i guess it is because `D` language not in it's queries https://github.com/andymass/vim-matchup#tree-sitter-integration (bad implementation relying on queries). Have not tested it in supported languages.
Bottlenecks: `synstack()`, `searchpair()` and relying on treesitter queries.

matchparen.nvim (with debounce = 0):
- syntax off (line 31): Always responsive.
- regex syntax (line 27): starts lagging around the half of the line. Right highlights inside strings.
- treesitter (line 31): Always responsive.
Bottlenecks: `synstack()`, relying on linear search in regex syntax highlighted buffers.

For `longline.d` file all implementations except matchparen.nvim behaves almost the same:
-syntax off: `matchparen.nvim` always responsive, others lagging on lines 21 and 23 (this PR for some reason lags on line 22 col 18 (space char)).
-syntax on (depends on `synmaxcol` option): all implementations starts lagging with much lower `synmaxcol` than `matchparen.nvim`.
-treesitter: `matchparen.nvim` always responsive and highlights. vimscript and new lua  matchparens lags and do not  highlight brackets on lines 21 and 23. This PR lags on lines 21 and 23 but highlights. vim-matchup behaves the same as `syntax off` because of missing query.

Bottleneck for this file is `synmaxcol`, `synstack()`, `searchpair()` and treesitter logic in plugins not skipping nodes and using linear search.

You can test it yourself (I have somewhat old PC, so your results can differ).
With debounce enabled matchparen.nvim feels almost always responsive. It's somewhat extreme examples you not often edit longline files, or files with so much visible brackets, but when you do it really annoying that cursor movement starts lagging because of the one plugin.

For me if you accepted reimplementation of matchparen plugin without `matchit` (for some reqson clason opinion was this https://github.com/neovim/neovim/issues/34658) then it would be better to adapt matchparen.nvim for it instead of relying on broken (wrong highlights) and slow old implementation.

justinmk commented:
@monkoose please post that in a issue so we can track it. And improvements are welcome so PR welcome, though the reimplementation should match the codebase conventions

monkoose commented:
@justinmk which issue?

clason commented:
New _and focused_ issue. Don't cherry-pick bad tree-sitter parsers; we want to improve the happy path, not cater to every bad grammar out there. (The goal is not to upstream a complex plugin but to -- even slightly -- improve the default experience. And the focus for core is always primarily on basic composable API, not bundled plugins.)
